patches/tModLoader/Terraria/WorldGen.cs.patch

```diff
@@ -1543,15 +1551,43 @@
  			AllowedToSpreadInfections = true;
  			CreativePowers.StopBiomeSpreadPower power = CreativePowerManager.Instance.GetPower<CreativePowers.StopBiomeSpreadPower>();
  			if (power != null && power.GetIsUnlocked())
-@@ -45669,7 +_,7 @@
+@@ -45623,7 +_,7 @@
+ 				Liquid.skipCount = 0;
+ 			}
+ 
+-			int worldUpdateRate = GetWorldUpdateRate();
++			double worldUpdateRate = GetWorldUpdateRate();
+ 			if (worldUpdateRate == 0)
  				return;
  
- 			for (int l = 0; l < Main.dayRate; l++) {
+@@ -45668,8 +_,14 @@
+ 			if (Main.dayTime)
+ 				return;
+ 
++			timePass += Main.desiredWorldEventsUpdateRate;
++			if (timePass < 1)
++				return;
++
++			double timePassReference = timePass;
+-			for (int l = 0; l < Main.dayRate; l++) {
++			for (int l = 0; l < (int)timePassReference; l++) {
++				timePass -= 1.0;
 -				float num6 = Main.maxTilesX / 4200;
-+				float num6 = Main.maxTilesX / 4200f; // Selfish fix for falling stars on extra small worlds
++				float num6 = Main.maxTilesX / 4200f;
  				num6 *= Star.starfallBoost;
  				if (!((float)Main.rand.Next(8000) < 10f * num6))
  					continue;
+@@ -45702,8 +_,8 @@
+ 			}
+ 		}
+ 
+-		public static int GetWorldUpdateRate() {
++		public static double GetWorldUpdateRate() {
+-			int result = Math.Min(Main.desiredWorldTilesUpdateRate, 24);
++			double result = Math.Min(Main.desiredWorldTilesUpdateRate, 24);
+ 			if (CreativePowerManager.Instance.GetPower<CreativePowers.FreezeTime>().Enabled)
```

Comment:
In all the places where you have 1E-38, consider investigating how freezeTime works, and using that instead.